### PR TITLE
FilterIntervals now filters out any singleton intervals

### PR DESCRIPTION
**Pull request by asmirnov239**
This implements suggestion #2 proposed by @samuelklee from issue #5852

### CODE REVIEW

**Comment by asmirnov239** [comment]
@samuelklee @mwalker174 Could you please review? Also does either of you know if ScatterIntervals task in CNVTasks WDL has a potential of producing singleton intervals?

**Comment by samuelklee** [comment]
@asmirnov239 looks like you are getting an NPE---remember that intervals are resolved after argument validation, so you need to do the check later.  Perhaps change the check to a filtering step in GermlineCNVCaller?

**Comment by asmirnov239** [comment]
@samuelklee Thank you, that's fixed now. 

Regarding  ScatterIntervals issue - what do you think about IntervalListTools logic to not produce singletons instead of filtering them out in GermlineCNVCaller?

**Comment by samuelklee** [comment]
@asmirnov239 I think that we probably want all shard sizes to be as close to equal as possible, so that we avoid any possible issues arising from different model capacities across shards.

Actually, is this even an issue at the GermlineCNVCaller step?  Maybe we just need to worry about it during postprocessing?

**Comment by asmirnov239** [comment]
@samuelklee You're right - I tested it and it's only an issue in postprocessing , so I'm only making a check there. Neither GermlineCNVCaller or DetermineContigPloidy code fails due to this issue.

It should be all set to be reviewed now.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervals.java [hunk 1]
```diff
@@ -431,6 +432,17 @@ private SimpleIntervalCollection filterIntervals() {
                     countNumberPassing(mask), numIntersectedIntervals));
         }
 
+        //finally, filter intervals that are solitary in their corresponding contigs
+        final Map<String, Long> contigToCountMap = IntStream.range(0, numIntersectedIntervals)
+                .filter(i -> !mask[i]).mapToObj(i -> intersectedIntervals.getRecords().get(i))
+                .collect(Collectors.groupingBy(SimpleInterval::getContig, Collectors.counting()));
+        IntStream.range(0, numIntersectedIntervals).filter(i -> !mask[i]).forEach(i -> {
+            final long value = contigToCountMap.get(intersectedIntervals.getRecords().get(i).getContig());
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervals.java:440]: I might just name this variable `count`.  Perhaps even `intervalCount`, and rename the map to `contigToIntervalCountMap`.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervals.java [hunk 2]
```diff
@@ -431,6 +432,17 @@ private SimpleIntervalCollection filterIntervals() {
...
+            if (value <= 1) {
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervals.java:441]: Seems a bit weird to have `<=` since it should never happen (and I can't imagine future code changes that would make it possible), but no biggie.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervals.java [hunk 3]
```diff
@@ -431,6 +432,17 @@ private SimpleIntervalCollection filterIntervals() {
...
+                mask[i] = true;
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervals.java:442]: Perhaps a log message here indicating when a contig is dropped?

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/PostprocessGermlineCNVCalls.java [hunk 4]
```diff
@@ -654,4 +653,25 @@ private static File getCopyNumberSegmentsFile(final File pythonSegmenterOutputPa
...
+     * Validate that the union of shard's interval lists does not have singleton intervals, i.e. intervals that
+     * are the only ones on their corresponding contigs.
+     */
+    private void checkForSingletonIntervalAbsence(final List<SimpleIntervalCollection> intervalCollections){
+        final List<SimpleInterval> concatenatedIntervalList = new ArrayList<>();
+        IntStream.range(0, numShards).forEach(i -> {
+            concatenatedIntervalList.addAll(intervalCollections.get(i).getIntervals());
+        });
+
+        final Map<String, Long> contigToCountMap = concatenatedIntervalList.stream()
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/PostprocessGermlineCNVCalls.java:667]: I'd make the variable names and code here consistent with the code in FilterIntervals.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/PostprocessGermlineCNVCalls.java [hunk 5]
```diff
@@ -654,4 +653,25 @@ private static File getCopyNumberSegmentsFile(final File pythonSegmenterOutputPa
         }
         return unsortedIntervalCollectionsFromModels;
     }
+
+    /**
+     * Validate that the union of shard's interval lists does not have singleton intervals, i.e. intervals that
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/PostprocessGermlineCNVCalls.java:658]: `Validate that the concatenation of the sharded interval lists does not have singleton intervals...`, perhaps?

File: src/test/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervalsIntegrationTest.java [hunk 6]
```diff
@@ -27,6 +27,7 @@
 import org.testng.annotations.Test;
 
 import java.io.File;
+import java.lang.reflect.AnnotatedElement;
```
Comment (samuelklee) [inline src/test/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervalsIntegrationTest.java:30]: Stray import?

File: src/test/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervalsIntegrationTest.java [hunk 7]
```diff
@@ -71,6 +72,16 @@
                             new AnnotationMap(Arrays.asList(
                                     Pair.of(CopyNumberAnnotations.GC_CONTENT, 0.95),
                                     Pair.of(CopyNumberAnnotations.MAPPABILITY, 0.95),
+                                    Pair.of(CopyNumberAnnotations.SEGMENTAL_DUPLICATION_CONTENT, 0.5)))),
+                    new AnnotatedInterval(new SimpleInterval("20", 51,	60),
```
Comment (samuelklee) [inline src/test/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervalsIntegrationTest.java:76]: White space.  Thanks for adding these tests!

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervals.java [hunk 8]
```diff
@@ -40,15 +40,16 @@
  * {@link CollectReadCounts}, outputs a filtered Picard interval list.  The set intersection of intervals from the
  * specified intervals, the annotated intervals, and the first count file will be taken as the initial set of intervals
  * on which to perform filtering.  Parameters for filtering based on the annotations and counts can be adjusted.
- * Annotation-based filters will be applied first, followed by count-based filters.  The result may be passed via -L to
- * other tools (e.g., {@link DetermineGermlineContigPloidy} and {@link GermlineCNVCaller}) to mask intervals from
- * analysis.
+ * Annotation-based filters will be applied first, followed by count-based filters. In the end, any singleton intervals
+ * (i.e. those being by themselves on their corresponding contigs) found after applying other filters will be filtered
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervals.java:44]: Comma after `i.e.`.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervals.java (hunk 3 above)
Comment (mwalker174) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervals.java:442]: Log a warning here, something like:

```
Filtered singleton interval on contig: "chrX"
```

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/PostprocessGermlineCNVCalls.java [hunk 9]
```diff
@@ -246,7 +243,7 @@
     public void onStartup() {
         super.onStartup();
         /* check for successful import of gcnvkernel */
-        PythonScriptExecutor.checkPythonEnvironmentForPackage("gcnvkernel");
+        PythonScriptExecutor.checkPythonEnvironmentForPackage("numpy");
```
Comment (mwalker174) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/PostprocessGermlineCNVCalls.java:246]: Why this change?

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/PostprocessGermlineCNVCalls.java [hunk 10]
```diff
@@ -654,4 +653,25 @@ private static File getCopyNumberSegmentsFile(final File pythonSegmenterOutputPa
         }
         return unsortedIntervalCollectionsFromModels;
     }
+
+    /**
+     * Validate that the union of shard's interval lists does not have singleton intervals, i.e. intervals that
+     * are the only ones on their corresponding contigs.
+     */
+    private void checkForSingletonIntervalAbsence(final List<SimpleIntervalCollection> intervalCollections){
```
Comment (mwalker174) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/PostprocessGermlineCNVCalls.java:661]: I'd argue you can just call this `checkForSingletonInterval`

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/PostprocessGermlineCNVCalls.java [hunk 11]
```diff
@@ -654,4 +653,25 @@ private static File getCopyNumberSegmentsFile(final File pythonSegmenterOutputPa
...
+
+        final Map<String, Long> contigToCountMap = concatenatedIntervalList.stream()
+                .collect(Collectors.groupingBy(SimpleInterval::getContig, Collectors.counting()));
+        contigToCountMap.keySet().forEach(c -> {
+            if (contigToCountMap.get(c) == 1) {
+                throw new IllegalArgumentException(
+                        String.format("Records contain a singleton interval on contig (%s)." +
+                                " Please run FilterIntervals tool first.", c));
+            }
+        });
```
Comment (mwalker174) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/PostprocessGermlineCNVCalls.java:675]: Can be simplified a bit:
```
        intervalCollections.stream()
                .flatMap(list -> list.getIntervals().stream())
                .collect(Collectors.groupingBy(SimpleInterval::getContig, Collectors.counting()))
                .entrySet().stream()
                .forEach(entry -> { 
                    if (entry.getValue() == 1) { 
                        throw new IllegalArgumentException(
                                String.format("Records contain a singleton interval on contig (%s)." + 
                                        " Please run FilterIntervals tool first.", entry.getKey()));
            }
        });
```

File: src/test/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervalsIntegrationTest.java [hunk 12]
```diff
@@ -92,28 +103,28 @@
...
+                {intervalsWithExtraIntervalFile, Collections.emptyList(), annotatedIntervalsFile, 0.1, 0.9, 0.1, 0.9, 0., 1., Arrays.asList(1, 2, 5)},
+                {intervalsWithExtraIntervalFile, Collections.emptyList(), annotatedIntervalsFile, 0.1, 0.9, 0., 1., 0.1, 0.9, Arrays.asList(2, 5)},
+                {intervalsWithExtraIntervalFile, Collections.emptyList(), annotatedIntervalsFile, 0., 1., 0.1, 0.9, 0.1, 0.9, Arrays.asList(2, 3, 5)},
+                {intervalsWithExtraIntervalFile, Collections.emptyList(), annotatedIntervalsFile, 0.1, 0.9, 0.1, 0.9, 0.1, 0.9, Arrays.asList(2, 5)},
+                {intervalsWithExtraIntervalFile, Collections.singletonList("20:1-10"), annotatedIntervalsFile, 0., 1., 0., 1., 0., 1., Arrays.asList(1, 2, 3, 4, 5)},
+                {intervalsWithExtraIntervalFile, Arrays.asList("20:1-15", "20:35-45"), annotatedIntervalsFile, 0., 1., 0., 1., 0., 1., Arrays.asList(2, 5)},
+                {intervalsWithExtraIntervalFile, Collections.singletonList("20:25-50"), annotatedIntervalsFile, 0.1, 0.9, 0., 1., 0., 1., Arrays.asList(0, 1, 5)}};
```
Comment (mwalker174) [inline src/test/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervalsIntegrationTest.java:127]: Does one of these test for the singleton case?

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervals.java [hunk 13]
```diff
@@ -431,6 +432,17 @@ private SimpleIntervalCollection filterIntervals() {
                     countNumberPassing(mask), numIntersectedIntervals));
         }
 
+        //finally, filter intervals that are solitary in their corresponding contigs
+        final Map<String, Long> contigToCountMap = IntStream.range(0, numIntersectedIntervals)
+                .filter(i -> !mask[i]).mapToObj(i -> intersectedIntervals.getRecords().get(i))
```
[the remark this answers is not in the document]
Comment (asmirnov239) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervals.java:437]: Done

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervals.java [hunk 14]
```diff
@@ -431,6 +432,17 @@ private SimpleIntervalCollection filterIntervals() {
...
+                .collect(Collectors.groupingBy(SimpleInterval::getContig, Collectors.counting()));
+        IntStream.range(0, numIntersectedIntervals).filter(i -> !mask[i]).forEach(i -> {
```
[the remark this answers is not in the document]
Comment (asmirnov239) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervals.java:439]: Done

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervals.java (hunk 1 above)
Comment (asmirnov239) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervals.java:440]: Done and done

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervals.java (hunk 2 above)
Comment (asmirnov239) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervals.java:441]: Changed it to `==`

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervals.java (hunk 3 above)
Comment (asmirnov239) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervals.java:442]: Done!

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervals.java (hunk 8 above)
Comment (asmirnov239) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervals.java:44]: Done

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/PostprocessGermlineCNVCalls.java (hunk 4 above)
Comment (asmirnov239) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/PostprocessGermlineCNVCalls.java:667]: Okay I did that as much as possible

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/PostprocessGermlineCNVCalls.java (hunk 5 above)
Comment (asmirnov239) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/PostprocessGermlineCNVCalls.java:658]: Changed

File: src/test/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervalsIntegrationTest.java (hunk 6 above)
Comment (asmirnov239) [reply src/test/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervalsIntegrationTest.java:30]: Hmm not sure how that got in there. Thanks!

File: src/test/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervalsIntegrationTest.java (hunk 7 above)
Comment (asmirnov239) [reply src/test/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervalsIntegrationTest.java:76]: Oh the other ones had tabs as well.. fixed!

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervals.java (hunk 3 above)
Comment (asmirnov239) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervals.java:442]: Done!

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/PostprocessGermlineCNVCalls.java (hunk 9 above)
Comment (asmirnov239) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/PostprocessGermlineCNVCalls.java:246]: Oops, this was to get it to run in IDE. Thanks for catching this!

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/PostprocessGermlineCNVCalls.java (hunk 10 above)
Comment (asmirnov239) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/PostprocessGermlineCNVCalls.java:661]: Changed

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/PostprocessGermlineCNVCalls.java (hunk 11 above)
Comment (asmirnov239) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/PostprocessGermlineCNVCalls.java:675]: Nice, I used this

File: src/test/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervalsIntegrationTest.java (hunk 12 above)
Comment (asmirnov239) [reply src/test/java/org/broadinstitute/hellbender/tools/copynumber/FilterIntervalsIntegrationTest.java:127]: All of them actually do because I added an interval from contig "21" which should be filtered out in all of these.
